File: src/lib/data/saved.ts
import { getDb } from "@/lib/db";
import { demoPlaces } from "@/lib/data/demo-data";

export interface SavedPlaceRecord {
  id: string;
  placeId: string;
  placeName: string;
  placeAddress: string;
  createdAt: string;
}

interface DbSavedPlace {
  id: string;
  placeId: string;
  createdAt: Date;
  place?: { id: string; name: string; address: string | null; city: string } | null;
}

const demoStore = new Map<string, SavedPlaceRecord>();

function placeName(placeId: string): { name: string; address: string } | null {
  const place = demoPlaces.find((p) => p.id === placeId);
  return place ? { name: place.name, address: `${place.address}, ${place.city}` } : null;
}

export async function listSavedFor(userId: string): Promise<{
  data: SavedPlaceRecord[];
  source: string;
}> {
  const db = getDb();
  if (!db) {
    const data = [...demoStore.values()]
      .filter((record) => record.placeId)
      .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
    return { data, source: "demo" };
  }

  const rows = await db.savedPlace.findMany({
    where: { userId },
    include: { place: true },
    orderBy: { createdAt: "desc" },
  });
  return {
    data: rows.map(toRecord),
    source: "database",
  };
}

export async function toggleSaved(
  userId: string,
  placeId: string,
): Promise<{ data: SavedPlaceRecord | null; saved: boolean; source: string }> {
  const db = getDb();
  if (!db) {
    const key = `${userId}:${placeId}`;
    const existing = demoStore.get(key);
    if (existing) {
      demoStore.delete(key);
      return { data: null, saved: false, source: "demo" };
    }
    const place = placeName(placeId);
    if (!place) return { data: null, saved: false, source: "demo" };
    const record: SavedPlaceRecord = {
      id: key,
      placeId,
      placeName: place.name,
      placeAddress: place.address,
      createdAt: new Date().toISOString(),
    };
    demoStore.set(key, record);
    return { data: record, saved: true, source: "demo" };
  }

  const existing = await db.savedPlace.findUnique({
    where: { userId_placeId: { userId, placeId } },
  });
  if (existing) {
    await db.savedPlace.delete({ where: { userId_placeId: { userId, placeId } } });
    return { data: null, saved: false, source: "database" };
  }

  const created = await db.savedPlace.create({
    data: { userId, placeId },
    include: { place: true },
  });
  return { data: toRecord(created), saved: true, source: "database" };
}

export async function removeSaved(
  userId: string,
  placeId: string,
): Promise<{ saved: boolean; source: string }> {
  const db = getDb();
  if (!db) {
    const key = `${userId}:${placeId}`;
    demoStore.delete(key);
    return { saved: false, source: "demo" };
  }
  const existing = await db.savedPlace.findUnique({
    where: { userId_placeId: { userId, placeId } },
  });
  if (!existing) return { saved: false, source: "database" };
  await db.savedPlace.delete({ where: { userId_placeId: { userId, placeId } } });
  return { saved: false, source: "database" };
}

export async function isSaved(userId: string, placeId: string): Promise<boolean> {
  const db = getDb();
  if (!db) return demoStore.has(`${userId}:${placeId}`);
  const existing = await db.savedPlace.findUnique({
    where: { userId_placeId: { userId, placeId } },
  });
  return Boolean(existing);
}

function toRecord(row: DbSavedPlace): SavedPlaceRecord {
  return {
    id: row.id,
    placeId: row.placeId,
    placeName: row.place?.name ?? placeName(row.placeId)?.name ?? "Tempat",
    placeAddress: row.place
      ? `${row.place.address ?? ""}${row.place.address ? ", " : ""}${row.place.city}`
      : placeName(row.placeId)?.address ?? "",
    createdAt: row.createdAt.toISOString(),
  };
}

export function assertKnownPlace(placeId: string): void {
  const db = getDb();
  if (db) return;
  if (!demoPlaces.find((p) => p.id === placeId)) {
    throw new Error("PLACE_NOT_FOUND");
  }
}

export function assertKnownPlaceId(placeId: string): void {
  if (typeof placeId !== "string" || placeId.length === 0) {
    throw new Error("PLACE_INVALID");
  }
  assertKnownPlace(placeId);
}

File: backend/internal/store/saved.go
package store

import (
	"anapsi/backend/internal/database"
	"anapsi/backend/internal/util"
)

// SavedPlaceRecord is a saved-place row returned to the client.
type SavedPlaceRecord struct {
	ID           string `json:"id"`
	PlaceID      string `json:"placeId"`
	PlaceName    string `json:"placeName"`
	PlaceAddress string `json:"placeAddress"`
	CreatedAt    string `json:"createdAt"`
}

// IsSavedPlace reports whether a user has saved a place.
func IsSavedPlace(userID, placeID string) (bool, error) {
	var count int
	err := database.DB.QueryRow(`SELECT COUNT(*) FROM saved_places WHERE userId = ? AND placeId = ?`, userID, placeID).Scan(&count)
	if err != nil {
		return false, err
	}
	return count > 0, nil
}

// ListSavedFor returns the user's saved places, newest first.
func ListSavedFor(userID string) ([]SavedPlaceRecord, error) {
	rows, err := database.DB.Query(
		`SELECT s.id, s.placeId, COALESCE(p.name, 'Tempat'), COALESCE(CONCAT_WS(', ', p.address, p.city), ''), s.createdAt
		 FROM saved_places s
		 LEFT JOIN places p ON p.id = s.placeId
		 WHERE s.userId = ? ORDER BY s.createdAt DESC`,
		userID,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []SavedPlaceRecord
	for rows.Next() {
		var rec SavedPlaceRecord
		if err := rows.Scan(&rec.ID, &rec.PlaceID, &rec.PlaceName, &rec.PlaceAddress, &rec.CreatedAt); err != nil {
			continue
		}
		rec.CreatedAt = util.ToISO(rec.CreatedAt)
		out = append(out, rec)
	}
	return out, nil
}

// ToggleSaved adds or removes a saved place and reports the resulting state:
// `saved` is true when the place ended up saved, false when removed.
func ToggleSaved(userID, placeID string) (*SavedPlaceRecord, bool, error) {
	existing, err := IsSavedPlace(userID, placeID)
	if err != nil {
		return nil, false, err
	}
	if existing {
		if _, err := RemoveSaved(userID, placeID); err != nil {
			return nil, false, err
		}
		return nil, false, nil
	}
	id := "saved-" + util.RandID()
	_, err = database.DB.Exec(
		`INSERT INTO saved_places (id, userId, placeId, createdAt) VALUES (?, ?, ?, ?)`,
		id, userID, placeID, util.NowMySQL(),
	)
	if err != nil {
		return nil, false, err
	}
	var rec SavedPlaceRecord
	rec.ID = id
	rec.PlaceID = placeID
	_ = database.DB.QueryRow(
		`SELECT COALESCE(p.name, 'Tempat'), COALESCE(CONCAT_WS(', ', p.address, p.city), '') FROM places p WHERE p.id = ?`,
		placeID,
	).Scan(&rec.PlaceName, &rec.PlaceAddress)
	rec.CreatedAt = util.NowISO()
	return &rec, true, nil
}

// RemoveSaved deletes a saved place. The returned bool mirrors the old
// removeSaved contract: false means the place is no longer saved.
func RemoveSaved(userID, placeID string) (bool, error) {
	existing, err := IsSavedPlace(userID, placeID)
	if err != nil {
		return false, err
	}
	if !existing {
		return false, nil
	}
	if _, err := database.DB.Exec(`DELETE FROM saved_places WHERE userId = ? AND placeId = ?`, userID, placeID); err != nil {
		return false, err
	}
	return false, nil
}
